### PR TITLE
Middleware support

### PR DESCRIPTION
**Pull request by fridays**
Helps with:

Transform query object (https://github.com/fridays/next-routes/issues/42#issuecomment-305442498, https://github.com/fridays/next-routes/issues/49)
Custom render+cache integration (https://github.com/fridays/next-routes/issues/4, https://github.com/fridays/next-routes/issues/40, https://github.com/fridays/next-routes/pull/45)
Server-side redirects (https://github.com/fridays/next-routes/pull/22, https://github.com/fridays/next-routes/issues/41)

Middleware is one or many functions to process server-side, one by one, as long as they return a query object or promise.

Example:

```javascript
routes.add('hi', '/hi/:id', ({app, req, res, route, query}) => {
  // Return transformed query object
  return Object.assign(query, {some: 'thing'})

  // or promise
  return Promise.resolve(Object.assign(query, {some: 'thing'}))

  // or custom render
  renderAndCache(app, req, res, route.page, query)

  // or redirect
  res.writeHead(301, {'Location': `/hello/${query.id}`})
  res.end()
})

// Array of middleware functions
routes.add('hi', '/hi/:id', [one, two, three])
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,11 +28,34 @@ routes.add('about', '/about-us/:foo(bar|baz)', 'index')
 ```
 This file is used both on the server and the client.
 
-API: `routes.add(name, pattern, page = name)`
+API: `routes.add(name, pattern, page = name, middleware)`
 
 - `name` - The route name
 - `pattern` - Express-style route pattern (uses [path-to-regexp](https://github.com/pillarjs/path-to-regexp))
 - `page` - Page inside `./pages` to be rendered (defaults to `name`)
+- *optional:* `middleware` - One or many functions to process server-side, one by one, as long as they return a query object or promise.
```
Comment (HaNdTriX) [inline README.md]: Lets be consistent and add `*optional:*` to the `page` point, too.

Comment (fridays) [inline README.md]: Agreed. Let's do some more readme optimizations in another PR to keep it consistent and simple

**Comment by HaNdTriX** [comment]
After thinking a bit about this PR, I am not sure if the server only approach will introduce unexpected results.

Redirecting and query transformation should not be used because they don't work on client page transitions :flushed:. 

I the past I have used composed Higher-Order Components (as middleware) for these use cases.

**Comment by ajoslin** [comment]
A good way to add lots of flexibility to the library.

One gripe: We're getting too many arguments in the `route.add` function. How about we change to `route.add(page, path, options)`, with `options` allowing keys `{name, middleware}`?

With this addition more features could be added at a later point, and backwards compat can be achieved with `if (typeof options === 'string') page = options`.

**Comment by Sly777** [comment]
The code looks great but If I need to do this globally (not route by route) how can I do this?

We need to think about also how to use it globally. Maybe we need to add function like;

```
// i'm bad on naming :D
routes.onRoutePaired = ({app, req, res, route, query}) => {
  return Object.assign(query, {some: 'thing'})

  // or promise
  return Promise.resolve(Object.assign(query, {some: 'thing'}))

  // or custom render
  renderAndCache(app, req, res, route.page, query)

  // or redirect
  res.writeHead(301, {'Location': `/hello/${query.id}`})
  res.end()
};
```

**Comment by fridays** [comment]
@HaNdTriX You're right redirects and query transforms should go into getInitialProps where they run on both client and server. @ajoslin does that work in your case?

Actually since the routes file goes into the client bundle, let's forget this and continue with #45 which targets the right place for server handlers like `renderAndCache`
